### DISCUSSION TITLE
Not possible to `validate` `WebSocketRequest`?

### DISCUSSION THREAD

**Discussion by lmcd** [Alamofire/Alamofire#3851] [General]
I'm trying to figure out at what point a `WebSocketRequest` will throw an error if a bad response code comes back. It feels like step 1 would be to `validate` the request against a list of response codes, like any other request, but this functionality seems missing.

If, for example a server throws a 500 error, where would this be intercepted client-side?

**Comment by lmcd** [comment]
Ok I've noticed you can determine this in a `.completed` message, which doesn't seem ideal.
Also it seems like you have to fully wait 60 seconds for the connection to timeout before this is delivered. So if the server returns a 401 or something, to indicate the user is not authorised to access this WebSocket endpoint, the user must wait an entire minute for it to timeout.

**Comment by jshier** [comment]
For websockets, any failed upgrade is considered a failure, and then returned as part of the `completed` event, as you saw. If you want to see those events separately, you can use the `.validate()` method (I'm pretty sure I tested that). There's no 60s timeout on Alamofire's side, it should just end immediately. So that's either a server behavior or the failed upgrade isn't immediately completing the task locally.

I'm going to convert this to a discussion, we can open an issue for anything that needs to be fixed after investigation.

**Reply by lmcd** [reply]
I'm seeing no `validate` method on `WebSocketRequest`, and I'm seeing a huge delay before the first `completed` event came through.

The server hasn't changed at all, and I'm migrating from Vapor WebSocket which would return immediately. Will migrate back 😢

**Reply by jshier** [reply]
Ah true, the validators are implemented per subclass. 

I would appreciate it if you could investigate the cause of the delay, as my local testing shows invalid response codes fail the socket immediately, so it would be nice to know what's different here.

In the end, remember this is a preview feature and will have bugs and limitations. If you'd like to help find those bugs and limitations I'd appreciate it, but immediately migrating an existing codebase probably isn't a good idea.

**Reply by jshier** [reply]
For instance, if you set a breakpoint on line 369 in `WebSocketRequest.swift`, perhaps the listener is receiving an error when the task does not.

**Reply by jshier** [reply]
Yeah, looks like there's an initial failure callback in there I'm not handling (yet somehow retriers work?), so there will need to be something which hooks into that event to allow validation and failure of the request in general.

**Reply by jshier** [reply]
Ah, the task does complete, it's just not hooked up for the failure case.

**Comment by lmcd** [comment]
Also, because there is no validation, `responseCode` on `AFError` will return nil, even if there is an associated response code.

**Reply by jshier** [reply]
Yes, what error do you end up getting?

**Reply by jshier** [reply]
In a local test I end up getting the `.sessionTaskFailed` error, as the task does complete when there's a bad response. So I'm still not sure where the delay you're getting is coming from.

**Reply by lmcd** [reply]
It's a `.sessionTaskFailed` error. I feel like `responseCode` should be intelligent enough to extract the response code from an error that's not of a specific validation type.

**Reply by jshier** [reply]
And that's what Alamofire's `validation()` methods usually do, they just don't exist for websockets yet.

**Comment by jshier** [comment]
I've added #3853 to track the addition of `validate()`.

**Reply by lmcd** [reply]
I'm just adding it myself locally. Could I expect it to work as expected if I just copy all the extensions and the implementation from `DataRequest`, but just replacing `data` with `nil`?

**Reply by jshier** [reply]
For root storage, yes, though the actual `WebSocketRequest` would likely map that down to a new type that drops the `Data` altogether, since it doesn't seem possible to get the body data at all.

**Comment by lmcd** [comment]
I implemented this:


````
@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
extension WebSocketRequest {
        /// A closure used to validate a request that takes a URL request, a URL response and data, and returns whether the
    /// request was valid.
    public typealias Validation = (URLRequest?, HTTPURLResponse, Data?) -> ValidationResult

    /// Validates that the response has a status code in the specified sequence.
    ///
    /// If validation fails, subsequent calls to response handlers will have an associated error.
    ///
    /// - Parameter acceptableStatusCodes: `Sequence` of acceptable response status codes.
    ///
    /// - Returns:                         The instance.
    @discardableResult
    public func validate<S: Sequence>(statusCode acceptableStatusCodes: S) -> Self where S.Iterator.Element == Int {
        validate { [unowned self] _, response, _ in
            self.validate(statusCode: acceptableStatusCodes, response: response)
        }
    }

    /// Validates that the response has a content type in the specified sequence.
    ///
    /// If validation fails, subsequent calls to response handlers will have an associated error.
    ///
    /// - parameter contentType: The acceptable content types, which may specify wildcard types and/or subtypes.
    ///
    /// - returns: The request.
    @discardableResult
    public func validate<S: Sequence>(contentType acceptableContentTypes: @escaping @autoclosure () -> S) -> Self where S.Iterator.Element == String {
        validate { [unowned self] _, response, data in
            self.validate(contentType: acceptableContentTypes(), response: response, data: data)
        }
    }

    /// Validates that the response has a status code in the default acceptable range of 200...299, and that the content
    /// type matches any specified in the Accept HTTP header field.
    ///
    /// If validation fails, subsequent calls to response handlers will have an associated error.
    ///
    /// - returns: The request.
    @discardableResult
    public func validate() -> Self {
        let contentTypes: () -> [String] = { [unowned self] in
            acceptableContentTypes
        }
        return validate(statusCode: acceptableStatusCodes).validate(contentType: contentTypes())
    }
}
````

Add this:


````
    @discardableResult
    public func validate(_ validation: @escaping Validation) -> Self {
        let validator: () -> Void = { [unowned self] in
            guard error == nil, let response else { return }

            let result = validation(request, response, nil)

            if case let .failure(error) = result { self.error = error.asAFError(or: .responseValidationFailed(reason: .customValidationFailed(error: error))) }

            // eventMonitor?.request(self,
            //                       didValidateRequest: request,
            //                       response: response,
            //                       data: nil,
            //                       withResult: result)
        }

        validators.write { $0.append(validator) }

        return self
    }
````


Still doesn't wrap the error when `.completed` returns
